optuna/samplers/_nsga2.py

```diff
@@ -0,0 +1,340 @@
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "nsga2:generation"
+_PARENTS_KEY = "nsga2:parents"
+_POPULATION_CACHE_KEY_PREFIX = "nsga2:population"
+
+
+class NSGAIISampler(BaseSampler):
+    """Multi-objective sampler using the NSGA-II algorithm.
+
+    NSGA-II stands for "Nondominated Sorting Genetic Algorithm II",
+    which is a well known, fast and elitist multi-objective genetic algorithm.
+
+    For further information about NSGA-II, please refer to the following paper:
+
+    - `A fast and elitist multiobjective genetic algorithm: NSGA-II
+      <https://ieeexplore.ieee.org/document/996017>`_
+
+    Args:
+        population_size:
+            Number of individuals (trials) in a generation.
+
+        mutation_prob:
+            Probability of mutating each parameter when creating a new individual.
+            If :obj:`None` is specified, the value ``1.0 / len(parent_trial.params)`` is used
+            where ``parent_trial`` is the parent trial of the target individual.
+
+        crossover_prob:
+            Probability that a crossover (parameters swapping between parents) will occur
+            when creating a new individual.
+
+        swapping_prob:
+            Probability of swapping each parameter of the parents during crossover.
+
+        seed:
+            Seed for random number generator.
+
+    """
+
+    def __init__(
+        self,
+        population_size: int = 50,
+        mutation_prob: Optional[float] = None,
+        crossover_prob: float = 0.9,
+        swapping_prob: float = 0.5,
+        seed: Optional[int] = None,
+    ) -> None:
+        # TODO(ohta): Reconsider the default value of each parameter.
+
+        if not isinstance(population_size, int):
+            raise TypeError("`population_size` must be an integer value.")
+
+        if population_size < 2:
+            raise ValueError("`population_size` must be greater than or equal to 2.")
+
+        if not (mutation_prob is None or 0.0 <= mutation_prob <= 1.0):
+            raise ValueError(
+                "`mutation_prob` must be None or a float value within the range [0.0, 1.0]."
+            )
+
+        if not (0.0 <= crossover_prob <= 1.0):
+            raise ValueError("`crossover_prob` must be a float value within the range [0.0, 1.0].")
+
+        if not (0.0 <= swapping_prob <= 1.0):
+            raise ValueError("`swapping_prob` must be a float value within the range [0.0, 1.0].")
+
+        self._population_size = population_size
+        self._mutation_prob = mutation_prob
+        self._crossover_prob = crossover_prob
+        self._swapping_prob = swapping_prob
+        self._random_sampler = RandomSampler(seed=seed)
+        self._rng = np.random.RandomState(seed)
+
+    def reseed_rng(self) -> None:
+        self._random_sampler.reseed_rng()
+        self._rng = np.random.RandomState()
+
+    def infer_relative_search_space(
+        self,
+        study: Study,
+        trial: FrozenTrial,
```

Comment:
A nit to reduce the number of lines.
```suggestion
        self, study: Study, trial: FrozenTrial
```